db/reqlog.c

```diff
@@ -2754,3 +2765,153 @@ void reqlog_set_clnt(struct reqlogger *logger, struct sqlclntstate *clnt)
+            reqlog_logf(options->logger, REQL_INFO,
+                        "host=%s task=%s fp=%s count=%"PRId64" cost=%"PRId64" time=%"PRId64 " rows=%"PRId64"\n",
+                        options->st->host, options->st->task,
+                        fingerprint,
+                        cnt->count - cnt->last_count,
+                        cnt->cost - cnt->last_cost,
+                        cnt->timems - cnt->last_timems,
+                        cnt->rows - cnt->last_rows,
+                        cnt);
+            cnt->last_count = cnt->count;
+            cnt->last_cost = cnt->cost;
+            cnt->last_rows = cnt->rows;
+            cnt->last_timems = cnt->timems;
+        }
+    }
+    else {
+        reqlog_logf(options->logger, REQL_INFO, "host=%s task=%s fp=%s count=%"PRId64" cost=%"PRId64" time=%"PRId64 " rows=%"PRId64"\n", options->st->host, options->st->task, fingerprint, cnt->count, cnt->cost, cnt->timems, cnt->rows);
+    }
+    return 0;
+}
+
+static int dump_client_sql_data_single(void *ent, void *arg) {
+    nodestats_t *st = (nodestats_t*) ent;
+    struct dump_client_sql_options *options = (struct dump_client_sql_options*) arg;
+    options->st = st;
+    Pthread_mutex_lock(&st->rawtotals.lk);
+    if (st->rawtotals.fingerprints)
+        hash_for(st->rawtotals.fingerprints, dump_client_fingerprint, arg);
+    Pthread_mutex_unlock(&st->rawtotals.lk);
+    return 0;
+}
+
+void dump_client_sql_data(struct reqlogger *logger, int do_snapshot) {
+    struct dump_client_sql_options options = { .do_snap = do_snapshot, .logger = logger };
+    Pthread_rwlock_wrlock(&clientstats_lk);
+    if (clientstats)
+        hash_for(clientstats, dump_client_sql_data_single, &options);
+    Pthread_rwlock_unlock(&clientstats_lk);
+}
+
+void add_fingerprint_to_rawstats(struct rawnodestats *stats, char *fingerprint, int cost, int rows, int timems) {
+    Pthread_mutex_lock(&stats->lk);
+    if (stats->fingerprints == NULL) {
+        // TODO: where does this get destroyed?
+        stats->fingerprints = hash_init_o(offsetof(struct query_count, fingerprint), FINGERPRINTSZ);
+        if (stats->fingerprints == NULL)
+            abort();
+    }
+    struct query_count *ct = hash_find(stats->fingerprints, fingerprint);
+    if (ct == NULL) {
+        ct = calloc(1, sizeof(struct query_count));
+        memcpy(ct->fingerprint, fingerprint, FINGERPRINTSZ);
+        hash_add(stats->fingerprints, ct);
+    }
+    ct->count++;
+    ct->rows += rows;
+    ct->cost += cost;
+    ct->timems += timems;
+    Pthread_mutex_unlock(&stats->lk);
+}
+
+struct client_sql_systable_options {
+    struct client_sql_systable_data *stats;
+    int nstats;
+    int nalloced;
+
+    nodestats_t *st; 
+};
+
+int gather_client_sql_data_fingerprint(void *ent, void *arg) {
+    struct client_sql_systable_options *opt = (struct client_sql_systable_options*) arg;
+    struct query_count *cnt = (struct query_count*) ent;
+
+    if (opt->nstats == opt->nalloced) {
```

Comment:
Consider running once to get count?

Reply:
Would need to grab all the locks for all connections to keep the count stable - this is cheaper.